### ISSUE TITLE
Difficulty Importing/Using Pandoc

### ISSUE THREAD

**Issue by moss-xyz** [boisgera/pandoc#72]
Bit of an odd one here, didn't see anyone else with a similar issue that had been posted.

I seem to be able to *import* the python package, but unable to *use* it, meaning that the following fails:

```python
import pandoc

text = "Hello world!"

p = pandoc.read(text)
```

It generates the following exception (`CommandNotFound`):
```
plumbum.commands.processes.CommandNotFound: (CommandNotFound(...), 'pandoc', [<LocalPath C:\WINDOWS\system32>, <LocalPath C:\WINDOWS>, <LocalPath C:\WINDOWS\System32\Wbem>, <LocalPath C:\WINDOWS\System32\WindowsPowerShell\v1.0>, <LocalPath C:\WINDOWS\System32\OpenSSH>, ... (list of PATHs)]
```

Which in turns triggers the following error:
```
RuntimeError: cannot find the pandoc program.
paths:['C:\\WINDOWS\\system32', 'C:\\WINDOWS', 'C:\\WINDOWS\\System32\\Wbem', 'C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0', 'C:\\WINDOWS\\System32\\OpenSSH', ... (list of PATHs)]
```

I would have thought the solution to this was to specify the path to `pandoc.exe` with `pandoc.configure()`, but that doesn't work either:
```
pandoc.configure("path/to/pandoc.exe") # same error
```

Is `pandoc` expected to be in my PATH? I installed it from conda using the instructions in [the documentation](https://boisgera.github.io/pandoc/getting-started/), so not sure why this error is happening. I also tried to install an older version (3.2.1) with conda but it didn't work either.

**Comment by boisgera** [comment]
Hi @moss-xyz ,

No, `pandoc` does not need to be in your PATH (even if it makes things simpler).

Try `pandoc.configure(path="path/to/pandoc.exe")` or `pandoc.configure(False, "path/to/pandoc.exe")`.

The path is not the first but the second argument in configure (the first being "auto").

Makes me wonder if I should make this configure function keyword arguments only ...

Cheers,

Sébastien

**Comment by boisgera** [comment]
Ah, too bad, that would have been an easy fix ...

AFAICT, the error you have can only happen when the configuration flag `auto` is set to `True` (precisely [here](https://github.com/boisgera/pandoc/blob/master/src/pandoc/__init__.py#L90)). 

That's pretty weird ... except if you have triggered an automatic configuration *before* you explicitly configure the path.

For example, 

```python
import pandoc
import pandoc.types 

pandoc.configure(path="...")
```

would fail at `import pandoc.types` since the types hierarchy is generated at runtime and needs to know what version of pandoc you are using (and therefore, calls implicitly `pandoc.configure(auto=True)` unless you have already configured pandoc manually).

Do you still have the error if all you do is

```python
import pandoc
pandoc.configure(path="...")
```

?

**Comment by moss-xyz** [comment]
Yes - just running `import pandoc` followed by `pandoc.configure(path="...")` triggers the issue. I wasn't previously importing `pandoc.types` (did not think it was necessary?), but trying to import that triggers the same error as `pandoc.configure(path="...")`, which is the same error from the OP, as I say above.

**Comment by boisgera** [comment]
Can you invoke your pandoc executable via `plumbum`?

For example, my executable is located at "/home/boisgera/.pixi/bin/pandoc" and I can do:

```pycon
>>> import plumbum
>>> p = plumbum.local["/home/boisgera/.pixi/bin/pandoc"]
>>> print(p("--version"))
pandoc 3.2
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: /home/boisgera/.local/share/pandoc
Copyright (C) 2006-2024 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
```

**Comment by moss-xyz** [comment]
Fascinating! That does seem to work?

```py
>>> import plumbum
>>> p = plumbum.local["path/to/env/Library/bin/pandoc.exe"]
>>> print(p("--version"))
pandoc.exe 3.2.1
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: C:\Users\my-name\AppData\Roaming\pandoc
Copyright (C) 2006-2024 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
```

Curious what this will tell you

**Comment by boisgera** [comment]
Well, AFAICT the issue is not that the executable is not available through plumbum when you give it the proper path, but for some reason `configure` believes that you at requesting `auto = True` and then searches for pandoc in your PATH environnement variable and then fails to find it.

Unfortunately, I don't have access to Windows (not even via VirtualBox because of [ongoing technical issues at Microsoft](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/)).

Are you by any chance able to use a debugger to go step by step and see if we are executing the `auto=True` branch and why?

**Comment by boisgera** [comment]
> Unfortunately, I don't have access to Windows (not even via VirtualBox because of [ongoing technical issues at Microsoft](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/)).
> 
> Are you by any chance able to use a debugger to go step by step and see if we are executing the `auto=True` branch and why?

Nevermind, I can reproduce on Linux, I'll deal with this. Thanks for your feedback!

**Comment by moss-xyz** [comment]
Ah, great, let me know!

I've actually never used a debugger before but I had just gotten it set up and was stepping through the functions - very cool feature!

**Comment by boisgera** [comment]
There is a faulty logic here where we import `types` before we update the configuration. That's an issue since the import triggers a configuration with `auto=True` if the `_configuration` is still `None`.

```python
    if not read_only:  # set the configuration, update pandoc.types

        try:
            from . import types
        except ImportError:  # only sensible explanation:
            # the types module is actually being imported (interpreted)
            # and is calling configure.
            types = sys.modules["pandoc.types"]

        _configuration = {
            "auto": auto,
            "path": path,
            "version": version,
            "pandoc_types_version": pandoc_types_version,
        }

        types.make_types()
```

The code in `types.py` also has an issue. AFAICT if the configuration is defined, there is no need to recall configure.

```python
_configuration = pandoc.configure(read=True)
if _configuration is None:
    pandoc.configure(auto=True)
else:
    pandoc.configure(**_configuration)
```
